### DISCUSSION TITLE
Azurite EventGrid Emulation

### DISCUSSION THREAD

**Discussion by swissarmykirpan** [microsoft/aspire#5895] [Q&A]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I want to emulate EventGrid specifically with Azurite

### Describe the solution you'd like

Azurite writes logs whenever an operation occurs e.g `PUT /devstoreaccount1/test/costings-uk-prod.csv?se=2024-10-22T02%3A49%3A07Z&sig=GIB%2BWjwK5ITiGi345QNpnsd3nDqZhX6%2Bb7uvAca7Wmc%3D&sp=rwl&sr=c&sv=2018-03-28 HTTP/1.1`

What I want to do is watch the logs and then publish events based on what operations have occurred.

Can you recommend an approach to fetching the logs.

**Comment by davidfowl** [comment]
You want to read the logs of the storage container and then do what with them?

**Reply by swissarmykirpan** [reply]
Want to make a manual call to the Function with the EventGridTrigger

// Default URL for triggering event grid function in the local environment.
// http://localhost:7071/runtime/webhooks/EventGrid?functionName={functionname}

**Reply by swissarmykirpan** [reply]
I made an attempt at accessing the logs using the mechanism in the dashboard.

```c#
internal sealed class GetAzureStorageResourceLogsLifecycleHook : IDistributedApplicationLifecycleHook
{
    private readonly ResourceNotificationService _resourceNotificationService;
    private readonly ResourceLoggerService _resourceLoggerService;
    private readonly ILogger<GetAzureStorageResourceLogsLifecycleHook> _logger;
    private readonly HttpClient _httpClient;

    public GetAzureStorageResourceLogsLifecycleHook(
        ResourceNotificationService resourceNotificationService,
        ResourceLoggerService resourceLoggerService,
        IHttpClientFactory httpClientFactory,
        ILogger<GetAzureStorageResourceLogsLifecycleHook> logger)
    {
        _resourceNotificationService = resourceNotificationService;
        _resourceLoggerService = resourceLoggerService;
        _logger = logger;
        _httpClient = httpClientFactory.CreateClient();
        _httpClient.BaseAddress = new Uri("http://localhost:7071/runtime/webhooks/");
    }

    public async Task AfterResourcesCreatedAsync(DistributedApplicationModel appModel, CancellationToken cancellationToken = default)
    {
        await foreach (var notification in _resourceNotificationService.WatchAsync(cancellationToken))
        {
            if (notification.Resource.Name == "storage")
            {
                await foreach (var batch in _resourceLoggerService.WatchAsync(notification.ResourceId).WithCancellation(cancellationToken))
                {
                    ...
                }
            }
        }
    }
```

However it does not work consistently. Frequently it only logs one thing: `info: GetAzureStorageResourceLogsLifecycleHook[0]
      LogLine { LineNumber = 1, Content = 2024-10-02T21:57:27.7859301Z Skipping storage because it is not configured to be provisioned., IsErrorMessage = False }` and then nothing subsequent to that

**Reply by swissarmykirpan** [reply]
Had this in place for a week now. Seems to be stable since upgrading to 9 rc1. @davidfowl just wanted your opinion on whether this is a sensible solution.

**Reply by davidfowl** [reply]
Maybe look at one of these https://github.com/search?q=eventgrid+emulator&type=repositories
